### DISCUSSION TITLE
JPA multi-module

### DISCUSSION THREAD

**Discussion by masdem** [micronaut-projects/micronaut-data#1537] [General]
JPA fail to scan my second module where are locate my entities

when I  run the app, i got some error like
 `Entities not found for JPA configuration: 'default' within packages [com.mydomaine.entity]. Check that you have correctly specified a package containing JPA entities within the "jpa.default.entity-scan.packages" property in your application configuration and that those entities are either compiled with Micronaut or a build time index produced with @Introspected(packages="foo.bar", includedAnnotations=Entity.class) declared on your Application class`

In my `application.yml`
`jpa.default.entity-scan.packages: 'com.mydomaine.entity`

**Comment by dstepanov** [comment]
Check if you have the annotation processor added for the second method.

**Reply by masdem** [reply]
How to do that

**Reply by dstepanov** [reply]
Show me what dependencies you have there

**Reply by masdem** [reply]
In entity module
```
dependencies {
    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.6.10")
    implementation("javax.persistence:javax.persistence-api:2.2")
    implementation("io.micronaut.data:micronaut-data-hibernate-jpa:3.3.0")
    implementation("org.mindrot:jbcrypt:0.4")

}
```

**Reply by masdem** [reply]
And in my main module
```
dependencies {
    implementation(project(":entity-module"))
    kapt("io.micronaut:micronaut-http-validation")
    kapt("io.micronaut.data:micronaut-data-processor")
    kapt("io.micronaut.security:micronaut-security-annotations")
    implementation("io.micronaut:micronaut-http-client")
    implementation("io.micronaut:micronaut-jackson-databind")
    implementation("io.micronaut:micronaut-runtime")
    implementation("io.micronaut.data:micronaut-data-jdbc")
    implementation("io.micronaut.kotlin:micronaut-kotlin-runtime")
    implementation("io.micronaut.security:micronaut-security-jwt")
    implementation("io.micronaut.reactor:micronaut-reactor")
    implementation("io.micronaut.reactor:micronaut-reactor-http-client")
    implementation("io.micronaut.sql:micronaut-jdbc-hikari")
    implementation("jakarta.annotation:jakarta.annotation-api")
    implementation("org.jetbrains.kotlin:kotlin-reflect:${kotlinVersion}")
    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8:${kotlinVersion}")
    runtimeOnly("ch.qos.logback:logback-classic")
    runtimeOnly("org.postgresql:postgresql")
    testImplementation("org.testcontainers:junit-jupiter")
    testImplementation("org.testcontainers:postgresql")
    testImplementation("org.testcontainers:testcontainers")
    implementation("io.micronaut:micronaut-validation")

    runtimeOnly("com.fasterxml.jackson.module:jackson-module-kotlin")

}
```

**Reply by dstepanov** [reply]
I think you need to add ‘micronaut-inject’ to kapt

**Reply by masdem** [reply]
I added `kapt("io.micronaut:micronaut-inject-java")` and run app, the error still here 
```
Entities not found for JPA configuration: 'default' within packages [com.mydomaine.entity]. Check that you have correctly specified a package containing JPA entities within the "jpa.default.entity-scan.packages" property in your application configuration and that those entities are either compiled with Micronaut or a build time index produced with @Introspected(packages="foo.bar", includedAnnotations=Entity.class) declared on your Application class
```

**Reply by dstepanov** [reply]
Make sure it’s added to that entity module

**Reply by masdem** [reply]
In entity module
```
plugins {
    id("org.jetbrains.kotlin.jvm") version "1.6.10"
    id("org.jetbrains.kotlin.kapt") version "1.6.10"
}

dependencies {
    kapt("io.micronaut:micronaut-inject:3.4.4")
    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.6.10")
    implementation("javax.persistence:javax.persistence-api:2.2")
    implementation("io.micronaut.data:micronaut-data-hibernate-jpa:3.3.0")
    implementation("org.mindrot:jbcrypt:0.4")
}
```
I still have this error

**Reply by dstepanov** [reply]
Hmm try to add micronaut data processor

**Reply by masdem** [reply]
It’s added in my main module or you talk about to add in entity module

**Reply by dstepanov** [reply]
I'm talking about the entity module

**Reply by masdem** [reply]
I added it like this `kapt("io.micronaut.data:micronaut-data-processor:3.4.1")`  but I still have the same error

**Reply by dstepanov** [reply]
Can you please create a sample app? I can take a look what's wrong

**Reply by masdem** [reply]
I can push source code on github and share with you

**Reply by dstepanov** [reply]
Ok

**Reply by masdem** [reply]
This is project url on github: https://github.com/masdem/micronaut-multi-module-jpa

**Reply by dstepanov** [reply]
I have added `id("io.micronaut.library") version "3.3.0"` to the core module and the error disappeared.

**Reply by masdem** [reply]
Great!
You can do the push on master branch

**Reply by masdem** [reply]
IT'S WORK
THANK YOU!!!

**Comment by phamvanlinh20111993** [comment]
I have also the same error, when i running on Intelij it ok but when i buid to jar file and build: **java -jar ajt.jar** it throw an error 'Entities not found for JPA configuration: 'default' within packages', does something i miss? i  using _Micronaut 3.8.9 + Groovy + Maven_ and strictly follow this document: https://guides.micronaut.io/latest/micronaut-jpa-hibernate-maven-java.html. Here is my pom.xml:


_<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>vmo.du5.com</groupId>
    <artifactId>ajt-recruiter-post</artifactId>
    <version>0.1</version>
    <packaging>${packaging}</packaging>

    <parent>
        <groupId>io.micronaut</groupId>
        <artifactId>micronaut-parent</artifactId>
        <version>3.8.9</version>
    </parent>

    <properties>
        <packaging>jar</packaging>
        <jdk.version>11</jdk.version>
        <release.version>11</release.version>
        <micronaut.version>3.8.9</micronaut.version>
        <micronaut.runtime>netty</micronaut.runtime>
        <groovyVersion>3.0.13</groovyVersion>
        <exec.mainClass>vmo.du5.com.Application</exec.mainClass>
    </properties>

    <repositories>
        <repository>
            <id>central</id>
            <url>https://repo.maven.apache.org/maven2</url>
        </repository>
    </repositories>

    
    <dependencies>
        <dependency>
            <groupId>io.micronaut</groupId>
            <artifactId>micronaut-inject-groovy</artifactId>
            <scope>provided</scope>
            <exclusions>
                <exclusion>
                    <groupId>org.slf4j</groupId>
                    <artifactId>slf4j-api</artifactId>
                </exclusion>
            </exclusions>
        </dependency>

        <dependency>
            <groupId>org.codehaus.groovy</groupId>
            <artifactId>groovy</artifactId>
            <version>${groovyVersion}</version>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut</groupId>
            <artifactId>micronaut-inject-java</artifactId>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut.data</groupId>
            <artifactId>micronaut-data-processor</artifactId>
            <scope>provided</scope>
            <exclusions>
                <exclusion>
                    <groupId>javax.transaction</groupId>
                    <artifactId>javax.transaction-api</artifactId>
                </exclusion>
            </exclusions>
        </dependency>

        <dependency>
            <groupId>io.micronaut</groupId>
            <artifactId>micronaut-http-validation</artifactId>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut</groupId>
            <artifactId>micronaut-validation</artifactId>
            <scope>compile</scope>
            <exclusions>
                <exclusion>
                    <groupId>javax.validation</groupId>
                    <artifactId>validation-api</artifactId>
                </exclusion>
            </exclusions>
        </dependency>

        <dependency>
            <groupId>io.micronaut.microstream</groupId>
            <artifactId>micronaut-microstream-annotations</artifactId>
            <scope>provided</scope>
        </dependency>
        
        <dependency>
            <groupId>io.micronaut.openapi</groupId>
            <artifactId>micronaut-openapi</artifactId>
            <scope>provided</scope>
            <exclusions>
                <exclusion>
                    <groupId>org.slf4j</groupId>
                    <artifactId>slf4j-nop</artifactId>
                </exclusion>
            </exclusions>
        </dependency>

        <dependency>
            <groupId>io.micronaut</groupId>
            <artifactId>micronaut-inject</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut</groupId>
            <artifactId>micronaut-http-server-netty</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut.beanvalidation</groupId>
            <artifactId>micronaut-hibernate-validator</artifactId>
            <scope>compile</scope>
            <exclusions>
                <exclusion>
                    <groupId>jakarta.validation</groupId>
                    <artifactId>jakarta.validation-api</artifactId>
                </exclusion>
            </exclusions>
        </dependency>

        <dependency>
            <groupId>io.micronaut.serde</groupId>
            <artifactId>micronaut-serde-jackson</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut.groovy</groupId>
            <artifactId>micronaut-runtime-groovy</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut.microstream</groupId>
            <artifactId>micronaut-microstream</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>io.swagger.core.v3</groupId>
            <artifactId>swagger-annotations</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>jakarta.annotation</groupId>
            <artifactId>jakarta.annotation-api</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-classic</artifactId>
            <scope>runtime</scope>

        </dependency>

        
        <dependency>
            <groupId>io.micronaut.rxjava3</groupId>
            <artifactId>micronaut-rxjava3</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut.rxjava3</groupId>
            <artifactId>micronaut-rxjava3-http-client</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut.microstream</groupId>
            <artifactId>micronaut-microstream-rest</artifactId>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>io.micronaut.test</groupId>
            <artifactId>micronaut-test-spock</artifactId>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>org.spockframework</groupId>
            <artifactId>spock-core</artifactId>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>org.codehaus.groovy</groupId>
                    <artifactId>groovy-all</artifactId>
                </exclusion>
            </exclusions>
        </dependency>

        
        <dependency>
            <groupId>io.micronaut.sql</groupId>
            <artifactId>micronaut-hibernate-jpa</artifactId>
            <scope>compile</scope>
            <exclusions>
                <exclusion>
                    <groupId>javax.persistence</groupId>
                    <artifactId>javax.persistence-api</artifactId>
                </exclusion>
            </exclusions>
        </dependency>

        <dependency>
            <groupId>io.micronaut.sql</groupId>
            <artifactId>micronaut-jdbc-hikari</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>com.h2database</groupId>
            <artifactId>h2</artifactId>
            <scope>compile</scope>
        </dependency>

        <dependency>
            <groupId>jakarta.persistence</groupId>
            <artifactId>jakarta.persistence-api</artifactId>
            <version>2.2.3</version>
            <scope>compile</scope>
        </dependency>

    </dependencies>

    <build>
        <finalName>ajt-recruiter-post-v1.0</finalName>
        <plugins>
            <plugin>
                <groupId>io.micronaut.build</groupId>
                <artifactId>micronaut-maven-plugin</artifactId>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <configuration>
                    <includes>
                        <include>**/*Spec.*</include>
                        <include>**/*Test.*</include>
                    </includes>
                </configuration>
            </plugin>

            <plugin>
                <groupId>org.codehaus.gmavenplus</groupId>
                <artifactId>gmavenplus-plugin</artifactId>
                <version>1.13.0</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>addSources</goal>
                            <goal>generateStubs</goal>
                            <goal>compile</goal>
                            <goal>removeStubs</goal>
                            <goal>addTestSources</goal>
                            <goal>generateTestStubs</goal>
                            <goal>compileTests</goal>
                            <goal>removeTestStubs</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <configuration>
                    <annotationProcessorPaths>
                        <path>
                            <groupId>io.micronaut.data</groupId>
                            <artifactId>micronaut-data-hibernate-jpa</artifactId>
                            <version>3.9.7</version>
                        </path>
                    </annotationProcessorPaths>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>properties-maven-plugin</artifactId>
                <version>1.0.0</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>set-system-properties</goal>
                        </goals>
                        <configuration>
                            <properties>
                                <property>
                                    <name>groovy.target.directory</name>
                                    <value>${project.build.directory}/classes</value>
                                </property>
                                <property>
                                    <name>groovy.parameters</name>
                                    <value>true</value>
                                </property>
                            </properties>
                        </configuration>
                    </execution>
                </executions>
            </plugin>

        </plugins>
    </build>

    <pluginRepositories>
    </pluginRepositories>
</project>_
Please help @dstepanov . Thanks

**Reply by linhpv-vmo** [reply]
anyone help ? :(

**Reply by graemerocher** [reply]
Maven? Provide a full project example of what doesn't work and we can have a look

**Reply by linhpv-vmo** [reply]
ok ok, sure, here is my repo https://github.com/linhpv-vmo/micronaut-example/tree/main @graemerocher

**Reply by linhpv-vmo** [reply]
when i build project to jar file and run it, the error will throw like this: `Message: Entities not found for JPA configuration: 'default' within packages [vmo.du5.com.entity,vmo.du5.com,vmo.du5.com.*,vmo.du5.com.]. Check that you have correctly specified a package containing JPA entities within the "jpa.default.entity-scan.packages" property in your application configuration and that those entities are either compiled with Micronaut or a build time index produced with @Introspected(packages="foo.bar", includedAnnotations=Entity.class) declared on your Application class
Path Taken: SessionFactory.buildHibernateSessionFactoryBuilder(SessionFactoryBuilder sessionFactoryBuilder) --> SessionFactory.buildHibernateSessionFactoryBuilder([SessionFactoryBuilder sessionFactoryBuilder]) --> SessionFactoryBuilder.buildHibernateSessionFactoryBuilder([Metadata metadata],JpaConfiguration jpaConfiguration,String name)
        at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1921)
        at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:249)
        at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3326)
        at io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3684)
        at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:341)
        at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:194)
        at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)
        at vmo.du5.com.Application.main(Application.groovy:35)
Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [org.hibernate.boot.SessionFactoryBuilder]`

**Reply by phamvanlinh20111993** [reply]
@graemerocher  can you help?

**Reply by graemerocher** [reply]
will look tomorrow

**Reply by linhpv-vmo** [reply]
how it's going Mr @graemerocher
